anesthetic/weighted_pandas.py

```diff
@@ -62,29 +79,48 @@ def compress(self, nsamples=None):
             compression). If <=0, then compress so that all weights are unity.
 
         """
-        i = compress_weights(self.w, self.u, nsamples)
+        i = compress_weights(self.w, self._u, nsamples)
         return self.repeat(i)
 
 
 class WeightedDataFrame(pandas.DataFrame):
     """Weighted version of pandas.DataFrame."""
 
-    _metadata = ['w', 'u']
+    _metadata = ['_w', '_u_']
 
     @property
     def _constructor_sliced(self):
         def __constructor_sliced(*args, **kwargs):
-            return WeightedSeries(*args, w=self.w, u=self.u, **kwargs)
+            series = WeightedSeries(*args, w=self._w, **kwargs)
+            series._u_ = self._u_
+            return series
         return __constructor_sliced
 
     @property
     def _constructor(self):
         return WeightedDataFrame
 
+    @property
+    def w(self):
+        """Sample weights."""
+        if self._w is not None:
+            return self._w[self.index]
+        else:
+            return None
+
+    @property
+    def _u(self):
+        """Random number for consistent compression."""
+        return self._u_[self.index]
+
     def __init__(self, *args, **kwargs):
```

Comment:
I think the `__init__` function is conventionally placed first in a class. Is there a reason you put it further down? I was a bit confused when I was looking for it and didn't find it right away...

Reply:
The latest push should resolve this